crates/external_services/src/file_storage/aws_s3.rs

```diff
@@ -0,0 +1,157 @@
+                "aws s3 region must not be empty".into(),
+            ))
+        })?;
+
+        when(self.bucket_name.is_default_or_empty(), || {
+            Err(ApplicationError::InvalidConfigurationValueError(
+                "aws s3 bucket name must not be empty".into(),
+            ))
+        })
+    }
+}
+
+/// AWS S3 file storage client.
+#[derive(Debug, Clone)]
+pub struct AwsFileStorageClient {
+    /// AWS S3 client
+    pub inner_client: Client,
+    /// The name of the AWS S3 bucket.
+    pub bucket_name: String,
+}
+
+impl AwsFileStorageClient {
+    /// Creates a new AWS S3 file storage client.
+    pub async fn new(config: &AwsFileStorageConfig) -> Self {
+        let region_provider = RegionProviderChain::first_try(Region::new(config.region.clone()));
+        let sdk_config = aws_config::from_env().region(region_provider).load().await;
+        println!("config_check {:?}", config);
+        Self {
+            inner_client: Client::new(&sdk_config),
+            bucket_name: config.bucket_name.clone(),
+        }
+    }
+
+    /// Uploads a file to AWS S3.
+    pub async fn upload_file_to_s3(
+        &self,
+        file_key: String,
+        file: Vec<u8>,
+    ) -> CustomResult<(), AwsS3StorageError> {
+        let bucket_name = self.bucket_name.clone();
+        let upload_res = self
+            .inner_client
+            .put_object()
+            .bucket(bucket_name)
+            .key(file_key.clone())
+            .body(file.into())
+            .send()
+            .await;
+        upload_res.map_err(AwsS3StorageError::UploadFailure)?;
+        Ok(())
+    }
+
+    /// Deletes a file from AWS S3.
+    pub async fn delete_file_from_s3(
+        &self,
+        file_key: String,
+    ) -> CustomResult<(), AwsS3StorageError> {
+        let bucket_name = self.bucket_name.clone();
+        let delete_res = self
+            .inner_client
+            .delete_object()
+            .bucket(bucket_name)
+            .key(file_key)
+            .send()
+            .await;
+        delete_res.map_err(AwsS3StorageError::DeleteFailure)?;
+        Ok(())
+    }
+
+    /// Retrieves a file from AWS S3.
+    pub async fn retrieve_file_from_s3(
+        &self,
+        file_key: String,
+    ) -> CustomResult<Vec<u8>, AwsS3StorageError> {
+        let bucket_name = self.bucket_name.clone();
+        let get_res = self
+            .inner_client
+            .get_object()
+            .bucket(bucket_name)
+            .key(file_key)
+            .send()
+            .await;
+        let mut object = get_res.map_err(AwsS3StorageError::RetrieveFailure)?;
+        let mut received_data: Vec<u8> = Vec::new();
+        while let Some(bytes) = object
+            .body
+            .try_next()
+            .await
+            .into_report()
+            .change_context(AwsS3StorageError::InvalidFileRetrieved)?
+        {
+            received_data.extend_from_slice(&bytes); // Collect the bytes in the Vec
+        }
```

Comment:
We should be able to call `bytes.collect()?.to_vec()` to obtain the underlying raw bytes. Refer to the S3 SDK docs for more information.